File: go/internal/markers/markers.go
// Package markers parses suppression comments that consumers attach in the
// comment block directly above a branch or return to opt out of an
// err-coverage rule. A marker is recognized on any line of that adjacent
// block and must carry a reason of at least MinReason characters after the
// colon.
package markers

import (
	"go/ast"
	"go/token"
	"strings"
)

// MinReason is the floor on a suppression marker's reason length after
// trimming (D009): non-empty was too cheap (`ok` / `todo` passed). No
// keyword bans - length is a structural nudge; substance is judged at review.
const MinReason = 10

type Kind int

const (
	NoReport Kind = iota
	ParseSkip
	NilReturn
	TestSkip
)

func (k Kind) String() string {
	switch k {
	case NoReport:
		return "no-report"
	case ParseSkip:
		return "parse-skip"
	case NilReturn:
		return "nil-return"
	case TestSkip:
		return "test-skip"
	}
	return ""
}

type Marker struct {
	Kind   Kind
	Reason string
	Pos    token.Pos
}

type Index struct {
	file   *token.File
	groups []group
}

// group records a comment block's last line and the marker nearest that end
// (closest to a node placed directly below), if the block carries one.
type group struct {
	lastLine  int
	marker    Marker
	hasMarker bool
}

func Build(fset *token.FileSet, f *ast.File) *Index {
	tf := fset.File(f.Pos())
	idx := &Index{file: tf}
	for _, cg := range f.Comments {
		g := group{lastLine: tf.Line(cg.End())}
		for _, c := range cg.List {
			if m, ok := parse(c); ok {
				g.marker, g.hasMarker = m, true // later comment wins: nearest the node
			}
		}
		idx.groups = append(idx.groups, g)
	}
	return idx
}

var prefixes = []struct {
	kind   Kind
	prefix string
}{
	{NoReport, "no-report:"},
	{ParseSkip, "parse-skip:"},
	{NilReturn, "nil-return:"},
	{TestSkip, "test-skip:"},
}

func parse(c *ast.Comment) (Marker, bool) {
	if !strings.HasPrefix(c.Text, "//") {
		return Marker{}, false
	}
	text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
	for _, p := range prefixes {
		if strings.HasPrefix(text, p.prefix) {
			reason := strings.TrimSpace(strings.TrimPrefix(text, p.prefix))
			if len(reason) < MinReason {
				return Marker{}, false
			}
			return Marker{Kind: p.kind, Reason: reason, Pos: c.Slash}, true
		}
	}
	return Marker{}, false
}

// Above returns the marker carried by the comment block directly above node.
// The marker may sit on any line of that block, not only the line immediately
// above, so a reason too long for one line can be followed by human context.
func (idx *Index) Above(node ast.Node) (Marker, bool) {
	if node == nil || idx.file == nil {
		return Marker{}, false
	}
	line := idx.file.Line(node.Pos())
	for _, g := range idx.groups {
		if g.lastLine == line-1 && g.hasMarker {
			return g.marker, true
		}
	}
	return Marker{}, false
}

File: py/tackbox/pyrules/markers.py
"""Suppression markers, ported from the Go/JS F1 marker-block engine.

A marker suppresses a finding when `# <prefix> <reason>` appears in the
contiguous comment block whose bottom line sits directly above the flagged
node, with a non-empty reason. "Block" = a run of comment lines on consecutive
rows; a blank line or code breaks it. The marker may sit on any line of that
block (a long reason can spill onto adjacent comment lines). Prefix is
`no-report:` (swallowed-exception rules) or `test-skip:` (TBX008); each prefix
gets its own index so the two suppression channels stay independent.
"""

from __future__ import annotations

import tokenize

NO_REPORT = "no-report:"
TEST_SKIP = "test-skip:"

# D009: a marker's reason must be at least this many chars after trimming.
# Non-empty was too cheap (`ok` / `todo` passed).
_MIN_REASON = 10


def _marker_reason_ok(comment: str, prefix: str) -> bool:
    """True iff `comment` (a `#...` token) carries `prefix` with a reason of at
    least _MIN_REASON chars after trimming (D009; parity with the Go/JS/Java
    marker parsers)."""
    text = comment.lstrip("#").strip()
    if not text.startswith(prefix):
        return False
    return len(text[len(prefix):].strip()) >= _MIN_REASON


def _is_standalone(tok: tokenize.TokenInfo) -> bool:
    """True iff only whitespace precedes the comment on its own line: a
    comment trailing code must never join or start a standalone block."""
    return tok.line[: tok.start[1]].strip() == ""


class MarkerIndex:
    """Bottom lines of comment blocks that carry a valid marker for `prefix`."""

    def __init__(
        self,
        file_tokens: list[tokenize.TokenInfo] | None,
        prefix: str = NO_REPORT,
    ):
        self._prefix = prefix
        self._suppress_bottoms: dict[int, int] = {}
        if file_tokens:
            self._build(file_tokens)

    def _build(self, tokens: list[tokenize.TokenInfo]) -> None:
        comments = sorted(
            (tok.start[0], tok.string)
            for tok in tokens
            if tok.type == tokenize.COMMENT and _is_standalone(tok)
        )
        block_rows: list[int] = []
        block_marked = 0
        for row, text in comments:
            if block_rows and row != block_rows[-1] + 1:
                self._flush(block_rows, block_marked)
                block_rows, block_marked = [], 0
            block_rows.append(row)
            if _marker_reason_ok(text, self._prefix):
                block_marked = row
        self._flush(block_rows, block_marked)

    def _flush(self, rows: list[int], marked: int) -> None:
        if rows and marked:
            self._suppress_bottoms[rows[-1]] = marked

    def above(self, node_line: int) -> int | None:
        """Return the marker's line when its block ends above `node_line`."""
        return self._suppress_bottoms.get(node_line - 1)
